### DISCUSSION TITLE
how to generate Magic DRC files from foundry supplied files

### DISCUSSION THREAD

**Discussion by gkamendje** [fossi-foundation/open-pdks#401] [Q&A]
@RTimothyEdwards I was wondering if there is an "easy" way to generate Magic DRC rules files out of files such as LEF files or ASSURA or Calibre DRC files which are generally supplied by the foundries. Any pointer to relevant docs/scripts will be more than appreciated.

**Comment by stineje** [comment]
HI,

Creating the tech files requires intimate knowledge the PDK.  You can try to output and use some of it, but it will need some hand holding imho.  But, great question.

All my best,

James

On Sep 16, 2023, at 12:38 PM, G-A. Kamendje ***@***.***> wrote:

﻿
CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe

@RTimothyEdwards<https://github.com/RTimothyEdwards> I was wondering if there is an "easy" way to generate Magic DRC rules files out of files such as LEF files or ASSURA or Calibre DRC files which are generally supplied by the foundries. Any pointer to relevant docs/scripts will be more than appreciated.

**Comment by RTimothyEdwards** [comment]
@gkamendje :  Generating rules from LEF files is pretty easy because the LEF file encodes only basic spacing and width rules.  Those can all be directly translated to `width`, `spacing`, and `widespacing` rules in magic.  I am not familiar with how Assura encodes DRC rules, but Calibre rules are largely boolean operations.  They somewhat closely match what in magic are called `cif-drc` rules.  I don't know how much of that could be directly translated between the two.

**Reply by gkamendje** [reply]
@RTimothyEdwards  I will have a look at the Magic documentation and see what I can come up with. My goal is to be able to contribute at translaton script to the OpenRoadFlowScript project. One of the main issue that new users face while trying to jump on the Opensource Tools wagon is certainly the lack of easy mechanism to translate close source PDK design rules into a format that is understood by either Magic or KLayout.

**Reply by slagernate** [reply]
@gkamendje any luck?

**Comment by RTimothyEdwards** [comment]
@gkamendje :  The basic underlying problem is that the OpenROAD flow scripts don't understand the open_pdks installations.  The Openlane (and Openlane2) scripts---which is basically an alternative method to the flow scripts which still uses OpenROAD---is based on the open_pdks installation and has integrated scripts that make use of Magic and Netgen for extraction and LVS.  However, if you're looking for PDK setup files for open source tools for any arbitrary foundry process, whether they be Magic or Klayout or other, it is not a trivial problem, and it is only a problem because foundries only support commercial tools.  The problem has been solved for the sky130 and gf180mcu processes because the foundries elected to open-source those processes.

That said, I would welcome any effort to create a script that could help ease the task of generating DRC and extraction rule decks for either Magic or KLayout.  Any partial solution is better than none.

**Comment by gkamendje** [comment]
Unfortunately no i have not got hhe time to look at it.
